node/actors/sync_blocks/src/peers/mod.rs

```diff
@@ -223,6 +233,16 @@ impl PeerStates {
                  ({last_contiguous_stored_block}) is lesser than the old value ({prev_contiguous_stored_block})"
             );
         }
+        let prev_first_stored_block = peer_state.first_stored_block;
+        if prev_first_stored_block < first_stored_block {
+            tracing::warn!(
+                %prev_first_stored_block,
+                %first_stored_block,
+                "Bogus state update from peer: first stored block increased; this shouldn't happen"
```

Comment:
I thought that we would expect peers to GC old blocks (i.e. drop the prefix of the block) rather than backfill them. As I see it, we would expect peers to have the first block specified (number + hash) in the config, so it is not like it would dynamically move backwards then. What is your plan?

Reply:
Sorry, I don't expect this warning to be long-term, it just corresponds to the fact that *now* there's no way for a node to prune history. So, if a node reports that its history is pruned, it's a reason to suspect a bug. In the longer term, we should at least support nodes pruning their history, and maybe also backfilling history (not quite sure on this; it *could* be the case that snapshot syncing could work that way).

Reply:
There was indeed a logic bug here in that `PeerState.first_stored_block` was updated *conditionally*. I've fixed it and added tests for a peer that prunes / backfills history.